MatrixTable.show errors if it tries to show more columns than in the MT

replicate:

```
hl.utils.range_matrix_table(1, 1).annotate_entries(x=5).show()
```

```
E           is.hail.utils.HailException: array index out of bounds: 1 / 1. IR: (ArrayRef
E             (GetField entries
E               (In Struct{row_idx:Int32,entries:Array[Struct{x:Int32}]} 1))
E             (I3 ...
```

The fix should also fix this problem:

```
In [2]: hl.utils.range_matrix_table(1, 1).show()
+---------+
| row_idx |
+---------+
|   int32 |
+---------+
|       0 |
+---------+
showing first 43 columns
```

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2502,7 +2502,7 @@ class MatrixTable(ExprContainer):
         def __str__(self):
             s = self.table_show.__str__()
             if self.displayed_n_cols != self.actual_n_cols:
-                s += f"showing the first { self.displayed_n_cols } columns of { self.actual_n_cols }"
+                s += f"showing the first { self.displayed_n_cols } of { self.actual_n_cols } columns"
             return s
 
         def __repr__(self):
@@ -2512,7 +2512,7 @@ class MatrixTable(ExprContainer):
             s = self.table_show._repr_html_()
             if self.displayed_n_cols != self.actual_n_cols:
                 s += '<p>'
-                s += f"showing the first { self.displayed_n_cols } columns of { self.actual_n_cols }"
+                s += f"showing the first { self.displayed_n_cols } of { self.actual_n_cols } columns"
                 s += '</p>\n'
             return s
 

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2500,7 +2500,7 @@ class MatrixTable(ExprContainer):
 
         def __str__(self):
             s = self.table_show.__str__()
-            s += f"showing first { self.n_cols } { plural('column', self.n_cols) }"
+            s += f"showing the first { self.n_cols } columns"
             return s
 
         def __repr__(self):
@@ -2509,7 +2509,7 @@ class MatrixTable(ExprContainer):
         def _repr_html_(self):
             s = self.table_show._repr_html_()
             s += '<p>'
-            s += f"showing first { self.n_cols } { plural('column', self.n_cols) }"
+            s += f"showing the first { self.n_cols } columns"
             s += '</p>\n'
             return s
 

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2494,13 +2494,15 @@ class MatrixTable(ExprContainer):
         Env.backend().execute(MatrixWrite(self._mir, writer))
 
     class _Show:
-        def __init__(self, table, n_rows, n_cols, width, truncate, types):
+        def __init__(self, table, n_rows, actual_n_cols, displayed_n_cols, width, truncate, types):
             self.table_show = table._show(n_rows, width, truncate, types)
-            self.n_cols = n_cols
+            self.actual_n_cols = actual_n_cols
+            self.displayed_n_cols = displayed_n_cols
 
         def __str__(self):
             s = self.table_show.__str__()
-            s += f"showing the first { self.n_cols } columns"
+            if self.displayed_n_cols != self.actual_n_cols:
+                s += f"showing the first { self.displayed_n_cols } columns of { self.actual_n_cols }"
             return s
 
         def __repr__(self):
@@ -2508,9 +2510,10 @@ class MatrixTable(ExprContainer):
 
         def _repr_html_(self):
             s = self.table_show._repr_html_()
-            s += '<p>'
-            s += f"showing the first { self.n_cols } columns"
-            s += '</p>\n'
+            if self.displayed_n_cols != self.actual_n_cols:
+                s += '<p>'
+                s += f"showing the first { self.displayed_n_cols } columns of { self.actual_n_cols }"
+                s += '</p>\n'
             return s
 
     @typecheck_method(n_rows=nullable(int),
@@ -2554,7 +2557,6 @@ class MatrixTable(ExprContainer):
                        for f, x in struct_expression.flatten().items())
 
         if n_cols is None:
-            characters_per_field = 10
             import shutil
             (characters, _) = shutil.get_terminal_size((80, 10))
             characters -= 6 # borders
@@ -2564,7 +2566,8 @@ class MatrixTable(ExprContainer):
                 characters -= estimate_size(self.row_value)
             characters = max(characters, 0)
             n_cols = characters // (estimate_size(self.entry) + 4) # 4 for the column index
-        n_cols = min(self.count_cols(), n_cols)
+        actual_n_cols = self.count_cols()
+        displayed_n_cols = min(actual_n_cols, n_cols)
 
         t = self.localize_entries('entries', 'cols')
         t = t.key_by()
@@ -2578,7 +2581,7 @@ class MatrixTable(ExprContainer):
                 handler = display
             except ImportError:
                 handler = print
-        handler(MatrixTable._Show(t, n_rows, n_cols, width, truncate, types))
+        handler(MatrixTable._Show(t, n_rows, actual_n_cols, displayed_n_cols, width, truncate, types))
 
     def globals_table(self) -> Table:
         """Returns a table with a single row with the globals of the matrix table.

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2564,6 +2564,7 @@ class MatrixTable(ExprContainer):
                 characters -= estimate_size(self.row_value)
             characters = max(characters, 0)
             n_cols = characters // (estimate_size(self.entry) + 4) # 4 for the column index
+        n_cols = min(self.count_cols(), n_cols)
 
         t = self.localize_entries('entries', 'cols')
         t = t.key_by()

--- a/hail/python/hail/matrixtable.py
+++ b/hail/python/hail/matrixtable.py
@@ -2574,7 +2574,7 @@ class MatrixTable(ExprContainer):
         t = t.select(
             **{f: t[f] for f in self.row_key},
             **{f: t[f] for f in self.row_value if include_row_fields},
-            **{str(i): t.entries[i] for i in range(0, n_cols)})
+            **{str(i): t.entries[i] for i in range(0, displayed_n_cols)})
         if handler is None:
             try:
                 from IPython.display import display

